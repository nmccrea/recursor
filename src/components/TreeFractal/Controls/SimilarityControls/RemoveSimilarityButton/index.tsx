import React from "react"
import { useDispatch } from "react-redux"
import { removeOne } from "../../../../../state/similarities/actions"
import { SimilarityId } from "../../../../../state/similarities/types"
import Button from "react-bootstrap/Button"

interface RemoveSimilarityButtonProps {
  similarityId: SimilarityId
}

const RemoveSimilarityButton = ({
  similarityId,
}: RemoveSimilarityButtonProps) => {
  const dispatch = useDispatch()
  return (
    <Button
      variant="danger"
      onClick={() => {
        dispatch(removeOne(similarityId))
      }}
    >
      Remove
    </Button>
  )
}

export default RemoveSimilarityButton
