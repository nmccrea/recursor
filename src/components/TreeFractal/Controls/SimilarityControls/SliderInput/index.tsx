import React from "react"
import { useSelector, useDispatch } from "react-redux"
import { SimilarityInputSelector } from "../../../../../state/similarities/selectors"
import { SimilarityInputActionCreator } from "../../../../../state/similarities/actions"
import {
  SimilarityNumericInputs,
  SimilarityNumericInputKey,
} from "../../../../../state/similarities/types"

export interface SliderInputProps<Key extends SimilarityNumericInputKey> {
  selector: SimilarityInputSelector<Key>
  actionCreator: SimilarityInputActionCreator<Key>
  label: string
  min: SimilarityNumericInputs[Key]
  max: SimilarityNumericInputs[Key]
  step: SimilarityNumericInputs[Key]
}

const SliderInput = <Key extends SimilarityNumericInputKey>({
  selector,
  actionCreator,
  label,
  min,
  max,
  step,
}: SliderInputProps<Key>) => {
  const value = useSelector(selector)
  const dispatch = useDispatch()

  return (
    <label>
      {label}

      <input
        type="range"
        value={value}
        min={min}
        max={max}
        step={step}
        onChange={(event) =>
          dispatch(actionCreator(Number(event.target.value)))
        }
      />
    </label>
  )
}

export default SliderInput
