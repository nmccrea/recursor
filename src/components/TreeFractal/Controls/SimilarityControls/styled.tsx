import styled from "styled-components"
import { screen } from "styles/utilities"

const Container = styled.div`
  display: flex;
  flex-direction: column;
  padding-right: 1rem;

  ${screen.large} {
    padding: 0.5rem 0;
    flex-direction: row;
    align-items: center;
    justify-content: space-between;
  }
`

const ControlElement = styled.div`
  padding: 0.5rem 0;

  & + div {
    border-top: 1px solid lightgray;
  }

  ${screen.large} {
    padding: 0 0.5rem;

    & + div {
      border-top: unset;
      border-left: 1px solid lightgray;
    }
  }
`

export { Container, ControlElement }
