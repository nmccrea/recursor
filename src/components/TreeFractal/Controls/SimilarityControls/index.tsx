import React from "react"
import {
  getTranslationSetterFor,
  getScaleSetterFor,
  getAngleSetterFor,
  getDepthSetterFor,
} from "../../../../state/similarities/actions"
import {
  getTranslationSelectorFor,
  getScaleSelectorFor,
  getAngleSelectorFor,
  getDepthSelectorFor,
} from "../../../../state/similarities/selectors"
import { SimilarityId } from "../../../../state/similarities/types"
import RemoveSimilarityButton from "./RemoveSimilarityButton"
import SliderInput from "./SliderInput"

interface SimilarityControlsProps {
  similarityId: SimilarityId
}

const SimilarityControls = ({ similarityId }: SimilarityControlsProps) => (
  <div style={{ display: "flex" }}>
    <p>{similarityId}</p>

    <RemoveSimilarityButton similarityId={similarityId} />

    <SliderInput
      selector={getTranslationSelectorFor(similarityId)}
      actionCreator={getTranslationSetterFor(similarityId)}
      label="Translation"
      min={0}
      max={1.5}
      step={0.05}
    />

    <SliderInput
      selector={getScaleSelectorFor(similarityId)}
      actionCreator={getScaleSetterFor(similarityId)}
      label="Scale"
      min={0}
      max={2}
      step={0.05}
    />

    <SliderInput
      selector={getAngleSelectorFor(similarityId)}
      actionCreator={getAngleSetterFor(similarityId)}
      label="Angle"
      // `min` and `max` must both be multiples of `step`
      min={-Math.round((2 * Math.PI) / 0.0001) * 0.0001}
      max={Math.round((2 * Math.PI) / 0.0001) * 0.0001}
      step={0.0001}
    />

    <SliderInput
      selector={getDepthSelectorFor(similarityId)}
      actionCreator={getDepthSetterFor(similarityId)}
      label="Depth"
      min={0}
      max={8}
      step={1}
    />
  </div>
)

export default SimilarityControls
