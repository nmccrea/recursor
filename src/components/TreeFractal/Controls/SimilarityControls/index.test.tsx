import React from "react"
import * as reactRedux from "react-redux"
import { render } from "@testing-library/react"
import SimilarityControls from "."

describe("`<SimilarityControls />`", () => {
  it("renders correctly", () => {
    jest.spyOn(reactRedux, "useSelector").mockReturnValue(0.4836)
    jest.spyOn(reactRedux, "useDispatch").mockReturnValue(jest.fn())

    const { container } = render(
      <SimilarityControls similarityId={"test/subject"} />
    )

    expect(container).toMatchSnapshot()
  })
})
