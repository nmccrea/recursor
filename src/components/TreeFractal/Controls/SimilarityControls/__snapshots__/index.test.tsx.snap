// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`<SimilarityControls />\` renders correctly 1`] = `
<div>
  <div
    style="display: flex;"
  >
    <p>
      test/subject
    </p>
    <button>
      Remove
    </button>
    <label>
      Translation
      <input
        max="1.5"
        min="0"
        step="0.05"
        type="range"
        value="0.4836"
      />
    </label>
    <label>
      Scale
      <input
        max="2"
        min="0"
        step="0.05"
        type="range"
        value="0.4836"
      />
    </label>
    <label>
      Angle
      <input
        max="6.2832"
        min="-6.2832"
        step="0.0001"
        type="range"
        value="0.4836"
      />
    </label>
    <label>
      Depth
      <input
        max="8"
        min="0"
        step="1"
        type="range"
        value="0.4836"
      />
    </label>
  </div>
</div>
`;
